File: go/single_consumer.go
// single_consumer.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
)

// MyMessage — структура, аналогичная producer.go
type MyMessage struct {
	MsgID int    `json:"msg_id"`
	Text  string `json:"text"`
}

func main() {
	// Конфигурация: auto.commit включён
	consumer, err := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers":  "127.0.0.1:9094,127.0.0.1:9095,127.0.0.1:9096",
		"group.id":           "single-message-consumer-group",
		"auto.offset.reset":  "earliest",
		"enable.auto.commit": true,
	})
	if err != nil {
		panic(err)
	}
	defer consumer.Close()

	// Подписка на топик
	err = consumer.SubscribeTopics([]string{"my_topic"}, nil)
	if err != nil {
		panic(err)
	}

	fmt.Println("SingleMessageConsumer started...")

	// Ловим Ctrl+C, чтобы корректно завершаться
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	run := true
	for run {
		select {
		case <-sigChan:
			fmt.Println("SingleMessageConsumer stopped by signal")
			run = false
		default:
			// Пытаемся считать сообщение (1 сек таймаут)
			msg, err := consumer.ReadMessage(-1)
			if err != nil {
				// Если ошибка (например, таймаут), просто пропускаем
				// ReadMessage(timeout) выбросит ошибку, если timeout != -1
				// Если timeout = -1, ждём бесконечно, пока появится сообщение
				continue
			}

			// Десериализация
			var data MyMessage
			err = json.Unmarshal(msg.Value, &data)
			if err != nil {
				fmt.Printf("[SingleConsumer] Deserialization error: %v\n", err)
				continue
			}

			// Обработка
			fmt.Printf("[SingleConsumer] Received message: Partition=%d Offset=%d Key=%s Value=%+v\n",
				msg.TopicPartition.Partition, msg.TopicPartition.Offset, string(msg.Key), data)
		}
	}

	fmt.Println("Closing consumer...")
}

File: python/single_consumer.py
# single_consumer.py
import json
import sys
from confluent_kafka import Consumer, KafkaError

def main():
    consumer_conf = {
        'bootstrap.servers': '127.0.0.1:9094,127.0.0.1:9095,127.0.0.1:9096',
        'group.id': 'single-message-consumer-group',
        'auto.offset.reset': 'earliest',      # Если нет сохранённого оффсета, начинать с начала топика
        'enable.auto.commit': True            # Автоматический коммит
    }

    consumer = Consumer(consumer_conf)
    consumer.subscribe(['my_topic'])

    print("SingleMessageConsumer started...")

    try:
        while True:
            msg = consumer.poll(timeout=1.0)
            if msg is None:
                continue  # нет сообщений – ждём дальше
            if msg.error():
                # Если пришла ошибка, просто логируем
                if msg.error().code() == KafkaError._PARTITION_EOF:
                    # Достигнут конец партиции
                    continue
                else:
                    print(f"[SingleConsumer] Error: {msg.error()}")
                    continue

            # Пробуем десериализовать
            try:
                data = json.loads(msg.value().decode('utf-8'))
                print(f"[SingleConsumer] Received message: {data}")
            except Exception as e:
                print(f"[SingleConsumer] Deserialization error: {e}", file=sys.stderr)

    except KeyboardInterrupt:
        print("SingleMessageConsumer stopped by user")
    finally:
        consumer.close()

if __name__ == '__main__':
    main()
